From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Tue, 29 Nov 2022 14:53:07 +1000
Subject: Xext: free the screen saver resource when replacing it
Origin: https://gitlab.freedesktop.org/xorg/xserver/commit/d6c7de9eadca980c8ce3b3b7752b67bfa95e6f31
Bug-Debian-Security: https://security-tracker.debian.org/tracker/CVE-2022-46343
Bug-Debian: https://bugs.debian.org/1026071

This fixes a use-after-free bug:

When a client first calls ScreenSaverSetAttributes(), a struct
ScreenSaverAttrRec is allocated and added to the client's
resources.

When the same client calls ScreenSaverSetAttributes() again, a new
struct ScreenSaverAttrRec is allocated, replacing the old struct. The
old struct was freed but not removed from the clients resources.

Later, when the client is destroyed the resource system invokes
ScreenSaverFreeAttr and attempts to clean up the already freed struct.

Fix this by letting the resource system free the old attrs instead.

CVE-2022-46343, ZDI-CAN 19404

This vulnerability was discovered by:
Jan-Niklas Sohn working with Trend Micro Zero Day Initiative

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
Acked-by: Olivier Fourdan <ofourdan@redhat.com>
(cherry picked from commit 842ca3ccef100ce010d1d8f5f6d6cc1915055900)
---
 Xext/saver.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: xorg-server-1.16.4/Xext/saver.c
===================================================================
--- xorg-server-1.16.4.orig/Xext/saver.c	2023-01-29 11:26:31.356082419 +0100
+++ xorg-server-1.16.4/Xext/saver.c	2023-01-29 11:26:31.352082432 +0100
@@ -1064,7 +1064,7 @@
         pVlist++;
     }
     if (pPriv->attr)
-        FreeScreenAttr(pPriv->attr);
+        FreeResource(pPriv->attr->resource, AttrType);
     pPriv->attr = pAttr;
     pAttr->resource = FakeClientID(client->index);
     if (!AddResource(pAttr->resource, AttrType, (void *) pAttr))
